### ISSUE TITLE
New Rule: Command substitution inside a shell?

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#3242]
#### For new checks and feature suggestions
- [X] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

* https://github.com/koalaman/shellcheck/issues?q=sort%3Aupdated-desc%20is%3Aissue%20command%20substitution%20unset%20variable
* https://github.com/koalaman/shellcheck/issues?q=sort%3Aupdated-desc%20is%3Aissue%20command%20substitution

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash
set -euo pipefail

determine() {
  # unbound var here is inside subshell; errexit is ignored ▶ subshell dies only
  tmp=$( true "$unset_var" )
  echo inside
}

# call in a subshell via command substitution
out=$( determine )
echo "returned: ›$out‹"
echo passed
```

#### Here's what shellcheck currently says:

```
$ shellcheck myscript
 
[Line 6:](javascript:setPosition(6, 3))
  tmp=$( true "$unset_var" )
  ^-- [SC2034](https://www.shellcheck.net/wiki/SC2034) (warning): tmp appears unused. Verify use (or export if used externally).
               ^-- [SC2154](https://www.shellcheck.net/wiki/SC2154) (warning): unset_var is referenced but not assigned.

$
```

_unrelated_

#### Here's what I wanted or expected to see:

```
$ shellcheck myscript
 
[Line 11:](javascript:setPosition(11, 3))
  out=$( determine )
  ^-- [SCXXXX](https://www.shellcheck.net/wiki/SCXXXX) (warning): Note that `determine` is NOT covered by the `set -u` guardrails. _See this https://mywiki.wooledge.org/BashFAQ/this article and/or use `set -E` + `trap 'exit 1' ERR`_

$
````

I know that the subject is complicated, and the warning above should've worked (if we are assuming one is CONSISTENTLY using lowercase env variables, so as for SC2154 to trigger) - but "a series of unfortunate events" lead me to hit this.

https://mywiki.wooledge.org/BashFAQ/112 (`set -u`) + https://mywiki.wooledge.org/BashFAQ/105 (`set -e`) do not seem to mention anything. Even if this feature cannot be completed - I am leaving this as a TIL-posterity.

**Keywords**: command substitution unset variable
